net-p2p/bitcoin-core/bitcoin-core-25.0.ebuild

```diff
@@ -0,0 +1,340 @@
+
+	sed -e 's/^\(complete -F _bitcoind\b\).*$/\1'"$(usev daemon ' bitcoind')$(usev qt5 ' bitcoin-qt')/" \
+		-i contrib/completions/bash/bitcoind.bash-completion || die
+}
+
+src_configure() {
+	local wallet ; if use berkdb || use sqlite ; then wallet=enable ; else wallet=disable ; fi
+	local myeconfargs=(
+		--disable-static
+		--${wallet}-wallet
+		$(use_with sqlite)
+		$(use_with berkdb bdb)
+		$(use_enable systemtap usdt)
+		$(use_with upnp miniupnpc)
+		$(use_with nat-pmp natpmp)
+		$(use_enable test tests)
+		--disable-bench
+		--disable-fuzz{,-binary}
+		$(use_with qrcode qrencode)
+		--disable-ccache
+		$(use_enable asm)
+		$(use_enable zeromq zmq)
+		$(use_enable man)
+		$(use_enable external-signer)
+		--with-utils
+		$(use_enable {bitcoin,util}-cli)
+		--enable-util-tx
+		--${wallet}-util-wallet
+		--disable-util-util
+		$(use_with libs)
+		$(use_with daemon)
+		$(use_with qt5 gui qt5)
+		$(use_with dbus qtdbus)
+		$(use_with system-leveldb)
+		$(use_with system-libsecp256k1)
+	)
+	econf "${myeconfargs[@]}"
+}
+
+src_install() {
+	use external-signer && DOCS+=( doc/external-signer.md )
+	use berkdb || use sqlite && DOCS+=( doc/managing-wallets.md )
+	use libs && DOCS+=( doc/shared-libraries.md )
+	use systemtap && DOCS+=( doc/tracing.md )
+	use zeromq && DOCS+=( doc/zmq.md )
+
+	default
+
+	! use libs || find "${ED}" -name '*.la' -delete || die
+	! use test || rm -f -- "${ED}"/usr/bin/test_bitcoin{,-qt} || die
+
+	newbashcomp contrib/completions/bash/bitcoin-tx.bash-completion bitcoin-tx
+	use bitcoin-cli && newbashcomp contrib/completions/bash/bitcoin-cli.bash-completion bitcoin-cli
+	use daemon || use qt5 && newbashcomp contrib/completions/bash/bitcoind.bash-completion bitcoind
+
+	if use daemon ; then
+		insinto /etc/bitcoin
+		sed -ne 's/ To use, copy this file$//p;Tp;:0;n;/save the file\.$/!b0;d;:p;p' \
+			share/examples/bitcoin.conf >"${ED}/etc/bitcoin/bitcoin.conf" || die
+		fowners bitcoin:bitcoin /etc/bitcoin/bitcoin.conf
+		fperms 0660 /etc/bitcoin/bitcoin.conf
+
+		newconfd contrib/init/bitcoind.openrcconf bitcoind
+		newinitd "${FILESDIR}/bitcoind.openrc" bitcoind
+		systemd_newunit contrib/init/bitcoind.service bitcoind.service
+
+		keepdir /var/lib/bitcoind
+		fperms 0750 /var/lib/bitcoind
+		fowners bitcoin:bitcoin /var/lib/bitcoind
+		dosym -r {/etc/bitcoin,/var/lib/bitcoind}/bitcoin.conf
+
+		insinto /etc/logrotate.d
+		newins "${FILESDIR}/bitcoind.logrotate-r1" bitcoind
+	fi
+
+	if use qt5 ; then
+		insinto /usr/share/icons/hicolor/scalable/apps
+		newins src/qt/res/src/bitcoin.svg bitcoin128.svg
+
+		domenu "${FILESDIR}/org.bitcoin.bitcoin-qt.desktop"
+
+		if use kde ; then
+			insinto /usr/share/kservices5
+			doins "${FILESDIR}/bitcoin-qt.protocol"
+		fi
+	fi
+
+	if use examples ; then
+		docinto examples
+		dodoc -r contrib/{linearize,qos}
+		use zeromq && dodoc -r contrib/zmq
+	fi
+}
+
+efmt() {
+	: ${1:?} ; local l ; while read -r l ; do "${!#}" "${l}" ; done < <(fmt "${@:1:$#-1}")
```

Comment:
I think I like `efmt`, even though I am not sure if I now 100% of that it does. `fmt` is used to reformat paragraphs. Using the `!#` event designator seems to obfuscate a bit what is going on, why not `do "${1}" "${l}"`. Using `: ${1:?}` to check if a parameter is set is certainly uncommon. Not that I dislike it, just saying.

I would welcome something a bit more explicit and readable. Especially considering that this may be something we could be factoring out in an eclass one day.


```bash
efmt() {
   local elogger, line
   logger="${1}"
   [[ $logger ]] || die "Must specify a logger"
   while read -r line; do
      "${logger}" "${line}"
   done < <(fmt)
}
```

Reply:
> I think I like `efmt`, even though I am not sure if I now 100% of that it does.

When the text to be logged contains variable/parameter expansions of unknown length, it is impossible to wrap the text statically at 75 columns (79 columns for the assumed terminal width, minus 4 columns for the line decorations). `fmt` reflows the text ideally and avoids the hassle of trying to get it just right in the ebuild.

> Using the `!#` event designator seems to obfuscate a bit what is going on

That has nothing to do with events. `${!name}` is a variable indirection that takes the value of the `name` variable as the name of the variable to be substituted. For instance, if `foo=bar` and `bar=baz`, then `${!foo}` will expand to `baz`. In my case, the variable being indirected through is the special shell variable `$#`, which holds the number of positional arguments. Thus, `${!#}` expands to the value of the _last_ positional argument given. There are other ways to achieve this, but this is the shortest I have seen.

> Using `: ${1:?}` to check if a parameter is set is certainly uncommon.

Again, this is the shortest way to achieve aborting on an unset required argument that I have seen. Note, it doesn't merely return a failure exit code from the function; it kills the whole shell. If `efmt` were to be hoisted into an eclass (or even into an EABI), it would certainly be better to call `die` in the ebuild-author-mistake case, but here I knew I wouldn't be misusing the function, so I wanted the shortest possible assertion.

> I would welcome something a bit more explicit and readable.

The example code you gave does not do the same thing. In particular, it will fail when options are passed to `efmt`, as the logging function to use is only in the first positional argument in the case when no options are passed. If you wanted something extra verbose (and less run-time efficient) but functionally equivalent, it would have to be something more like:

```bash
efmt() {
    local arg efunc line
    local -a opts
    for arg ; do
        [[ -v efunc ]] && opts+=( "${efunc}" )
        efunc="${arg}"
    done
    [[ -n "${efunc}" ]] || die 'Must specify a logging function'
    while read -r line ; do
        "${efunc}" "${line}"
    done < <(fmt "${opts[@]}")
}
```

I wanted an unobtrusive one-liner since I was just trying to meet the [requirements of the ebuild writing guide](https://devmanual.gentoo.org/ebuild-writing/messages/index.html) without making a big production out of it.